src/Workflow/MoveJatsAssets.php

```diff
@@ -0,0 +1,164 @@
+use function implode;
+use function in_array;
+use function Libero\ContentApiBundle\stream_hash;
+use function Libero\ContentStore\element_uri;
+use function Libero\ContentStore\parse_media_type;
+use function preg_match;
+use function sprintf;
+
+final class MoveJatsAssets implements EventSubscriberInterface
+{
+    private const HAS_MIMETYPE_ATTRIBUTE = [
+        'graphic',
+        'inline-graphic',
+        'inline-media',
+        'inline-supplementary-material',
+        'media',
+        'supplementary-material',
+    ];
+
+    private $client;
+    private $concurrency;
+    private $filesystem;
+    private $origin;
+    private $publicUri;
+
+    public function __construct(
+        string $origin,
+        string $publicUri,
+        FilesystemInterface $filesystem,
+        ClientInterface $client,
+        int $concurrency = 10
+    ) {
+        $this->origin = $origin;
+        $this->publicUri = $publicUri;
+        $this->filesystem = $filesystem;
+        $this->client = $client;
+        $this->concurrency = $concurrency;
+    }
+
+    public static function getSubscribedEvents() : array
+    {
+        return [
+            'workflow.libero.content_store.put.transition.manipulate' => 'onManipulate',
+        ];
+    }
+
+    public function onManipulate(Event $event) : void
+    {
+        /** @var PutTask $task */
+        $task = $event->getSubject();
+
+        each_limit_all(
+            $this->processAssets($this->findLinkedElements($task->getDocument()), $task),
+            $this->concurrency
+        )->wait();
+    }
+
+    /**
+     * @return iterable<Element>
+     */
+    private function findLinkedElements(Document $document) : iterable
+    {
+        $document->registerNamespace('jats', 'http://jats.nlm.nih.gov');
+        $document->registerNamespace('xlink', 'http://www.w3.org/1999/xlink');
+
+        yield from $document('//jats:article//jats:*[@xlink:href]');
+    }
+
+    private function processAssets(iterable $elements, PutTask $task) : iterable
+    {
+        foreach ($elements as $element) {
+            if ($this->shouldProcess($element)) {
+                yield $this->processAsset($element, $task);
+            }
+        }
+    }
+
+    private function processAsset(Element $asset, PutTask $task) : PromiseInterface
+    {
+        return new Coroutine(
+            function () use ($asset, $task) : iterable {
+                /** @var ResponseInterface $response */
+                $response = yield $this->client->requestAsync('GET', element_uri($asset));
+
+                $contentType = parse_media_type($response->getHeaderLine('Content-Type'));
+                /** @var resource $stream */
+                $stream = $response->getBody()->detach();
+                $path = $this->pathFor($task, $contentType, $stream);
+
+                $this->updateElement($asset, $contentType, $path);
+                $this->deployAsset($contentType, $stream, $path);
```

Comment:
the current `DEBUG` level logs are succinct:
```
app_1        | [2019-05-28 12:18:48] request.INFO: Matched route "PUT_items_id_versions_version". {"route":"PUT_items_id_versions_version","route_parameters":{"_route":"PUT_items_id_versions_version","_controller":"libero.content_api.service.item.put","id":"new-article","version":"1"},"request_uri":"http://localhost:8080/items/new-article/versions/1","method":"PUT"} []
app_1        | [2019-05-28 12:19:02] doctrine.DEBUG: "START TRANSACTION" [] []
app_1        | [2019-05-28 12:19:02] doctrine.DEBUG: SELECT item.current_version FROM items_items item WHERE item.id = :id FOR SHARE {"id":"new-article"} []
app_1        | [2019-05-28 12:19:02] doctrine.DEBUG: UPDATE items_versions SET content = ?, hash = ? WHERE id = ? AND version = ? ["[resource] (stream)","fa0548eae7f2f4e028dbc4f46e92c05a","new-article",1] []
app_1        | [2019-05-28 12:19:02] doctrine.DEBUG: "COMMIT" [] []
web_1        | 172.23.0.1 - - [28/May/2019:12:19:02 +0000] "PUT /items/new-article/versions/1 HTTP/1.1" 204 25 "-" "curl/7.47.0" "-"
app_1        | 172.23.0.5 -  28/May/2019:12:18:48 +0000 "PUT /index.php" 204
```
This function seems like a gold mine for `DEBUG` level logs about what is happening to which asset. Could also be elevated to `INFO` as this is part of a write API so needs some form of auditing/debugging.

Reply:
6208e4f3950bc6f9382d21789be6703e5adaecfc and 72f0cd391cd4e879c65e610718feb001a48a94e1 will help as it logs HTTP requests (both downloads and uploads), though currently at `debug`.

Reply:
(Can configure: https://github.com/csarrazi/CsaGuzzleBundle/blob/master/src/Resources/doc/configuration_reference.md)

Reply:
Guzzle and AWS SDK logs now look good:
```
app_1        | [2019-05-28 12:44:21] app.DEBUG: 3e816d0e28fe GuzzleHttp/6.3.3 curl/7.64.0 PHP/7.3.2 - [28/May/2019:12:44:21 +0000] "GET /test/rc.jpg HTTP/1.1" 200 5965717 [] []
app_1        | [2019-05-28 12:44:21] app.DEBUG: 3e816d0e28fe aws-sdk-php/3.93.2 Symfony/4.2.3 SYMOD/2.0.1 GuzzleHttp/6.3.3 curl/7.64.0 PHP/7.3.2 - [28/May/2019:12:44:21 +0000] "PUT /bucket/new-article/v1/b92a9bc45d6dadf3c4054c11ddcab67a.jpeg HTTP/1.1" 200 0 [] []
app_1        | [2019-05-28 12:44:21] app.DEBUG: 3e816d0e28fe GuzzleHttp/6.3.3 curl/7.64.0 PHP/7.3.2 - [28/May/2019:12:44:21 +0000] "GET /test/rc.jpg HTTP/1.1" 200 5965717 [] []
app_1        | [2019-05-28 12:44:21] app.DEBUG: 3e816d0e28fe aws-sdk-php/3.93.2 Symfony/4.2.3 SYMOD/2.0.1 GuzzleHttp/6.3.3 curl/7.64.0 PHP/7.3.2 - [28/May/2019:12:44:21 +0000] "PUT /bucket/new-article/v1/b92a9bc45d6dadf3c4054c11ddcab67a.jpeg HTTP/1.1" 200 0 [] []
```